sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,377 @@
+    #     class_list.append("warning")
+    # elif name == "error":
+    #     class_list.append("danger")
+    # elif name in ("important", "hint"):
+    #     class_list.append("tip")
+    # elif name == "todo":
+    #     class_list.append("info")
+
+    class CustomizedAdmonition(CustomAdmonitionDirective):
+        default_title = title
+        classes = class_list
+        optional_arguments = int(name not in admonitionlabels)
+        node_class = nodes.admonition if name != "todo" else sphinx.ext.todo.todo_node
+
+    return CustomizedAdmonition
+
+
+def on_builder_inited(app: Sphinx):
+    """register the directives for the custom admonitions and build the CSS."""
+    config = app.config
+    custom_admonitions: List[CustomAdmonitionConfig] = getattr(
+        config, "sphinx_immaterial_custom_admonitions"
+    )
+    setattr(app.builder.env, "sphinx_immaterial_custom_icons", {})
+    for admonition in custom_admonitions:
+
+        app.add_directive(
+            name=admonition.name,
+            cls=get_directive_class(
+                admonition.name, admonition.title, admonition.classes
+            ),
+            override=admonition.override,
+        )
+
+        # set variables for CSS template to match HTML output from generated directives
+        admonition.name = nodes.make_id(admonition.name)
+        if admonition.icon is not None:
+            admonition.icon = load_svg_into_builder_env(app.builder, admonition.icon)
+    setattr(app.builder.env, "sphinx_immaterial_custom_admonitions", custom_admonitions)
+
+
+def on_config_inited(app: Sphinx, config: Config):
+    """Add admonitions based on CSS classes inherited from mkdocs-material theme."""
+
+    # override the generic admonition directive
+    app.add_directive("admonition", get_directive_class("admonition", ""), True)
+
+    # generate directives for inherited admonitions from upstream CSS
+    if getattr(config, "sphinx_immaterial_generate_inherited_admonitions"):
+        for admonition in (
+            "abstract",
+            "info",
+            "success",
+            "question",
+            "failure",
+            "bug",
+            "example",
+            "quote",
+        ):
+            app.add_directive(
+                admonition,
+                get_directive_class(admonition, _(admonition.title())),
+            )
+    confval_name = "sphinx_immaterial_custom_admonitions"
+    # validate user defined config for custom admonitions
+    user_defined_admonitions = pydantic.parse_obj_as(
+        List[CustomAdmonitionConfig], getattr(config, confval_name)
+    )
+    setattr(config, confval_name, user_defined_admonitions)
+    user_defined_dir_names = [directive.name for directive in user_defined_admonitions]
+
+    # override the specific admonitions defined in sphinx and docutils
+    # these are the admonitions that have translated titles in sphinx.locale
+    for admonition, title in admonitionlabels.items():
```

Comment:
I think we should have a config option to disable this, in case it conflicts with another extension that wants to modify these admonitions.  Also for the `admonition` directive itself.

Reply:
It really shouldn't conflict with other extensions, but I do understand the prudence. The syntax is practically the same (except for the optional arg). The visitor is still monkeypatched regardless.

While I'm at it, I'll add another config option to disallow overriding the built-in/specific admonitions in sphinx/docutils.